lib/std/unicode.zig

```diff
@@ -712,6 +712,29 @@ pub fn utf8ToUtf16Le(utf16le: []u16, utf8: []const u8) !usize {
     return dest_i;
 }
 
+pub fn utf8ToUtf16LeWriter(writer: anytype, utf8: []const u8) !usize {
+    var src_i: usize = 0;
+    var bytes_written: usize = 0;
+    while (src_i < utf8.len) {
+        const n = utf8ByteSequenceLength(utf8[src_i]) catch return error.InvalidUtf8;
+        const next_src_i = src_i + n;
+        const codepoint = utf8Decode(utf8[src_i..next_src_i]) catch return error.InvalidUtf8;
+        if (codepoint < 0x10000) {
+            const short = @intCast(u16, codepoint);
+            try writer.writeIntLittle(u16, short);
+            bytes_written += 2;
+        } else {
+            const high = @intCast(u16, (codepoint - 0x10000) >> 10) + 0xD800;
+            const low = @intCast(u16, codepoint & 0x3FF) + 0xDC00;
+            try writer.writeIntLittle(u16, high);
+            try writer.writeIntLittle(u16, low);
+            bytes_written += 4;
+        }
+        src_i = next_src_i;
+    }
+    return bytes_written;
+}
```

Comment:
This is something I'm the most unsure about as it might be a little strange. I needed something that:

- Is able to return an error on overflowing the provided buffer
- And is able to use exactly the buffer provided (i.e. no over-allocations)

None of the existing utf8 -> utf16le conversion functions provide both of those:

- `utf8ToUtf16LeWithNull` uses an ArrayList so providing a `FixedBufferAllocator` to it can fail even if the converted string would fit in the buffer because it can over-allocate when resizing
- `utf8ToUtf16Le` asserts that the buffer is large enough

Using a `Writer` got me both of my requirements essentially for free, but I'm not sure if it's out-of-place here or not.